lib/asert_daa.py

```diff
@@ -0,0 +1,156 @@
+# Electron Cash - lightweight Bitcoin Cash client
+# Copyright (C) 2020 The Electron Cash Developers
+#
+# Permission is hereby granted, free of charge, to any person
+# obtaining a copy of this software and associated documentation files
+# (the "Software"), to deal in the Software without restriction,
+# including without limitation the rights to use, copy, modify, merge,
+# publish, distribute, sublicense, and/or sell copies of the Software,
+# and to permit persons to whom the Software is furnished to do so,
+# subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+# BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import os
+
+from collections import namedtuple
+from typing import Optional, Union
+
+from .util import print_error
+
+def bits_to_target(bits: int) -> int:
+    size = bits >> 24
+    assert size <= 0x1d
+
+    word = bits & 0x00ffffff
+    assert 0x8000 <= word <= 0x7fffff
+
+    if size <= 3:
+        return word >> (8 * (3 - size))
+    else:
+        return word << (8 * (size - 3))
+
+def _get_asert_activation_mtp():
+    """ Returns 1605441600 (Nov 15, 2020 12:00:00 UTC) or whatever override may
+    be set by the env variable ASERT_MTP """
+    default_mtp = 1605441600  # Nov 15, 2020 12:00:00 UTC
+    mtp = os.environ.get('ASERT_MTP', default_mtp)
+    try: mtp = int(mtp)
+    except: pass
+    if not isinstance(mtp, int) or mtp <= 1510600000:
+        print_error("Error: Environment variable ASERT_MTP ignored because it is invalid: {}".format(str(mtp)))
+        mtp = default_mtp
+    if mtp != default_mtp:
+        print_error("ASERT_MTP of {} will be used".format(mtp))
+    return mtp
+
+class Anchor(namedtuple("Anchor", "height bits prev_time")):
+    pass
+
+class ASERTDaa:
+    """ Parameters and methods for the ASERT DAA. Instances of these live in
+    networks.TestNet, networks.MainNet as part of the chain params. """
+
+    MTP_ACTIVATION_TIME = _get_asert_activation_mtp()  # Normally Nov. 15th, 2020 UTC 12:00:00
+
+    IDEAL_BLOCK_TIME = 10 * 60  # 10 mins
+    TAU = 2 * 24 * 3600  # for mainnet, testnet has 3600 (1 hour) half-life
+    # Integer implementation uses these for fixed point math
+    RBITS = 16  # number of bits after the radix for fixed-point math
```

Comment:
Dunno if you want to change anything from the bchn version, but I found this to be a bit ugly (might as well just hardcode these values 16 and 65536, there is no freedom to change them).

Reply:
I see what you are saying but -- this is in a way "Self-documenting" -- if we took out these constants and hard-coded them, it would make the code slightly harder to read, no?  We'd have to then pepper more comments in the main calculation function to explain the magic constants anyway, no?  Would you be ok with just leaving these be as-is?

Reply:
Yeah no prob. :+1: